Deriving `Reconcile` on a struct with a `MaybeMissing` leads to compilation error

I'm trying to use autosurgeon with `MaybeMissing` instead of `Option`.  I would expect the following code to work:

```
 use autosurgeon::{hydrate::MaybeMissing, Hydrate, Reconcile};

#[derive(Hydrate, Reconcile)]
struct S {
    s: MaybeMissing<String>,
}
```

But it leads to this compilation error:

```
error[E0277]: the trait bound `MaybeMissing<std::string::String>: Reconcile` is not satisfied
   --> src/main.rs:3:19
    |
3   | #[derive(Hydrate, Reconcile)]
    |                   ^^^^^^^^^ the trait `Reconcile` is not implemented for `MaybeMissing<std::string::String>`, which is required by `&MaybeMissing<std::string::String>: Reconcile`
4   | struct S {
5   |     s: MaybeMissing<String>,
    |     - required by a bound introduced by this call
    |
    = help: the following other types implement trait `Reconcile`:
              bool
              i8
              i16
              i32
              i64
              u8
              u16
              u32
            and 17 others
    = note: required for `&MaybeMissing<std::string::String>` to implement `Reconcile`
note: required by a bound in `autosurgeon::reconcile::MapReconciler::put`
   --> /home/matt/.cargo/registry/src/index.crates.io-6f17d22bba15001f/autosurgeon-0.8.3/src/reconcile.rs:114:15
    |
114 |     fn put<R: Reconcile, P: AsRef<str>>(&mut self, prop: P, value: R) -> Result<(), Self::Error>;
    |               ^^^^^^^^^ required by this bound in `MapReconciler::put`
```

I'm curious to learn whether this is expected to work or not.

P.S. if there is any available way to induce the compiler to tell me what the other types in the "and 17 others" error is, that'd be awesome to learn about too!

--- a/autosurgeon/src/hydrate.rs
+++ b/autosurgeon/src/hydrate.rs
@@ -298,7 +298,7 @@ impl<T> HydrateResultExt<Option<T>> for Result<Option<T>, HydrateError> {
 /// Note that this can be combined with `Option<T>` to get both behaviours
 ///
 /// ```rust
-/// # use autosurgeon::{Hydrate, hydrate::MaybeMissing};
+/// # use autosurgeon::{Hydrate, MaybeMissing};
 /// # use automerge::transaction::Transactable;
 /// let mut doc = automerge::AutoCommit::new();
 /// let name = MaybeMissing::<Option<String>>::hydrate(&doc, &automerge::ROOT, "name".into()).unwrap();
@@ -328,6 +328,17 @@ impl<T: Hydrate> Hydrate for MaybeMissing<T> {
         match doc.get(obj, &prop)? {
             None => Ok(Self::Missing),
             Some(_) => T::hydrate(doc, obj, prop).map(Self::Present),
+        }
+    }
+}
+
+impl<T: crate::Reconcile> crate::Reconcile for MaybeMissing<T> {
+    type Key<'a> = T::Key<'a>;
+
+    fn reconcile<R: crate::Reconciler>(&self, reconciler: R) -> Result<(), R::Error> {
+        match self {
+            Self::Missing => Ok(()),
+            Self::Present(val) => val.reconcile(reconciler),
         }
     }
 }

--- a/autosurgeon/src/lib.rs
+++ b/autosurgeon/src/lib.rs
@@ -459,7 +459,7 @@ mod doc;
 pub use doc::{Doc, ReadDoc};
 pub mod hydrate;
 #[doc(inline)]
-pub use hydrate::{hydrate, hydrate_path, hydrate_prop, Hydrate, HydrateError};
+pub use hydrate::{hydrate, hydrate_path, hydrate_prop, Hydrate, HydrateError, MaybeMissing};
 pub mod reconcile;
 #[doc(inline)]
 pub use reconcile::{

--- a/autosurgeon/src/reconcile.rs
+++ b/autosurgeon/src/reconcile.rs
@@ -1093,4 +1093,49 @@ mod tests {
             }}
         );
     }
+
+    struct Foo {
+        bar: crate::hydrate::MaybeMissing<String>,
+    }
+
+    impl crate::Reconcile for Foo {
+        type Key<'a> = NoKey;
+
+        fn reconcile<R: Reconciler>(&self, mut reconciler: R) -> Result<(), R::Error> {
+            let mut map = reconciler.map()?;
+            map.put("bar", &self.bar)?;
+            Ok(())
+        }
+    }
+
+    #[test]
+    fn reconcile_maybe_missing_present() {
+        let mut doc = automerge::AutoCommit::new();
+        reconcile(
+            &mut doc,
+            &Foo {
+                bar: crate::MaybeMissing::Present("baz".to_string()),
+            },
+        )
+        .unwrap();
+        let val = doc.get(&automerge::ROOT, "bar").unwrap().unwrap();
+        assert_eq!(
+            val.0,
+            automerge::Value::Scalar(std::borrow::Cow::Owned(ScalarValue::Str("baz".into())))
+        );
+    }
+
+    #[test]
+    fn reconcile_maybe_missing_not_present() {
+        let mut doc = automerge::AutoCommit::new();
+        reconcile(
+            &mut doc,
+            &Foo {
+                bar: crate::MaybeMissing::Missing,
+            },
+        )
+        .unwrap();
+        let val = doc.get(&automerge::ROOT, "bar").unwrap();
+        assert!(val.is_none());
+    }
 }

